File: br_operator/ledger_builder.py
"""Ledger event builder for BlackRoad OS Operator.

Python equivalent of lib/ledger-builder.ts for server-side event construction.

This ensures all mesh/agent/infra events have required fields and conform
to the Amundson governance contract.

Usage:
    from br_operator.ledger_builder import (
        build_mesh_connect_event,
        build_agent_invoke_event,
        build_operator_infra_event,
        validate_ledger_event,
    )

@amundson 0.1.0
@governor alice.governor.v1
@operator alexa.operator.v1
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from uuid import uuid4

from br_operator.models.policy import (
    LedgerLevel,
    PolicyEffect,
    PolicyEvaluateRequest,
    PolicyEvaluateResponse,
)
from br_operator.models.ledger import (
    Layer,
    LedgerActor,
    LedgerDecision,
    LedgerEventCreate,
)

# ============================================
# CONSTANTS
# ============================================

LANGUAGE_VERSION = "0.1.0"

MESH_POLICY_PACK = {
    "name": "mesh-policies",
    "version": "1.0.0",
}

INFRA_POLICY_PACK = {
    "name": "infra-policies",
    "version": "1.0.0",
}

# Ledger level precedence
LEDGER_LEVEL_RANK: Dict[LedgerLevel, int] = {
    LedgerLevel.NONE: 0,
    LedgerLevel.DECISION: 1,
    LedgerLevel.ACTION: 2,
    LedgerLevel.FULL: 3,
}


# ============================================
# HELPERS
# ============================================

def _resolve_level(required: LedgerLevel, override: Optional[LedgerLevel] = None) -> LedgerLevel:
    """Ensure actual ledger level is >= required level."""
    if override is None:
        return required

    if LEDGER_LEVEL_RANK[override] < LEDGER_LEVEL_RANK[required]:
        return required

    return override


def _resolve_layer(host: str) -> Layer:
    """Resolve the layer from the host."""
    if "gov." in host or ".systems" in host:
        return Layer.GOVERNANCE
    if "mesh." in host or "agents." in host:
        return Layer.MESH
    if "api.blackroad.io" in host:
        return Layer.GATEWAY
    if "db." in host or "infra." in host or "ops." in host:
        return Layer.INFRA
    return Layer.EXPERIENCE


def _resolve_policy_scope(host: str) -> str:
    """Resolve the policy scope from the host."""
    if host.startswith("mesh.") or host.endswith(".blackroad.network"):
        return "mesh.*"
    if host.startswith("agents."):
        return "agents.*"
    if host.startswith("infra.") or host.startswith("ops."):
        return "infra.*"
    if host.startswith("gov."):
        return "gov.api.*"
    if host.startswith("edu."):
        return "edu.*"
    if host.startswith("homework."):
        return "homework.*"
    return "app.*"


def _effect_to_decision(effect: PolicyEffect) -> LedgerDecision:
    """Convert PolicyEffect to LedgerDecision."""
    return LedgerDecision(effect.value)


# ============================================
# BUILDER OPTIONS
# ============================================

class LedgerEventOptions:
    """Options for building a ledger event."""

    def __init__(
        self,
        request: PolicyEvaluateRequest,
        response: PolicyEvaluateResponse,
        resource_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        ledger_level_override: Optional[LedgerLevel] = None,
        request_context: Optional[Dict[str, Any]] = None,
        response_summary: Optional[Dict[str, Any]] = None,
        intent_id: Optional[str] = None,
        sequence_num: int = 0,
        policy_pack_name: Optional[str] = None,
        policy_pack_version: Optional[str] = None,
    ):
        self.request = request
        self.response = response
        self.resource_id = resource_id
        self.metadata = metadata or {}
        self.ledger_level_override = ledger_level_override
        self.request_context = request_context
        self.response_summary = response_summary
        self.intent_id = intent_id
        self.sequence_num = sequence_num
        self.policy_pack_name = policy_pack_name
        self.policy_pack_version = policy_pack_version


# ============================================
# CORE BUILDER
# ============================================

def build_ledger_event(options: LedgerEventOptions) -> LedgerEventCreate:
    """Build a ledger event from a policy evaluation request/response pair.

    This is the canonical way to create ledger events. It ensures:
    1. All required fields are present
    2. Ledger level is >= required_ledger_level
    3. Location fields are consistent with the host
    4. Evidence from the request is preserved
    """
    request = options.request
    response = options.response

    host = request.context.request_metadata.host or "app.blackroad.io"
    service = request.context.request_metadata.service or "unknown"
    correlation_id = request.context.request_metadata.correlation_id or str(uuid4())

    ledger_level = _resolve_level(response.required_ledger_level, options.ledger_level_override)

    # Build metadata with pack info
    metadata = {
        **options.metadata,
        "language_version": LANGUAGE_VERSION,
    }

    if options.policy_pack_name:
        metadata["policy_pack"] = options.policy_pack_name
    if options.policy_pack_version:
        metadata["policy_pack_version"] = options.policy_pack_version

    return LedgerEventCreate(
        # Identity
        correlation_id=correlation_id,
        intent_id=options.intent_id,
        sequence_num=options.sequence_num,

        # Location
        layer=_resolve_layer(host),
        host=host,
        service=service,
        policy_scope=_resolve_policy_scope(host),

        # Actor
        actor=LedgerActor(
            user_id=request.subject.user_id,
            role=request.subject.role,
            agent_id=request.subject.attributes.get("agent_id") if request.subject.attributes else None,
            delegation_id=None,  # Set from claims if present
        ),

        # Action
        action=request.action,
        resource_type=request.resource.type,
        resource_id=options.resource_id or request.resource.id,

        # Decision
        decision=_effect_to_decision(response.decision),
        policy_id=response.policy_id,
        policy_version=response.policy_version,

        # Evidence
        asserted_facts=request.context.asserted_facts,
        fact_evidence=request.context.fact_evidence,
        claims=request.context.claims,

        # Metadata
        ledger_level=ledger_level,
        metadata=metadata,
        request_context=options.request_context if ledger_level == LedgerLevel.FULL else None,
        response_summary=options.response_summary if ledger_level == LedgerLevel.FULL else None,
    )


# ============================================
# MESH/AGENT HELPERS
# ============================================

def build_mesh_connect_event(
    request: PolicyEvaluateRequest,
    response: PolicyEvaluateResponse,
    connection_id: str,
    connection_type: str = "websocket",
    client_info: Optional[Dict[str, Any]] = None,
) -> LedgerEventCreate:
    """Build a ledger event for mesh connection."""
    return build_ledger_event(LedgerEventOptions(
        request=request,
        response=response,
        resource_id=connection_id,
        metadata={
            "connection_type": connection_type,
            "client_info": client_info or {},
            "event_type": "mesh:connected",
        },
        policy_pack_name=MESH_POLICY_PACK["name"],
        policy_pack_version=MESH_POLICY_PACK["version"],
        request_context={
            "connection_id": connection_id,
            "connection_type": connection_type,
        },
        response_summary={
            "connected": response.decision in (PolicyEffect.ALLOW, PolicyEffect.WARN, PolicyEffect.SHADOW_DENY),
        },
    ))


def build_mesh_route_event(
    request: PolicyEvaluateRequest,
    response: PolicyEvaluateResponse,
    message_id: str,
    source: str,
    destination: str,
) -> LedgerEventCreate:
    """Build a ledger event for mesh message routing."""
    return build_ledger_event(LedgerEventOptions(
        request=request,
        response=response,
        resource_id=message_id,
        metadata={
            "source": source,
            "destination": destination,
            "event_type": "mesh:routed",
        },
        policy_pack_name=MESH_POLICY_PACK["name"],
        policy_pack_version=MESH_POLICY_PACK["version"],
    ))


def build_agent_invoke_event(
    request: PolicyEvaluateRequest,
    response: PolicyEvaluateResponse,
    invocation_id: str,
    agent_id: str,
    capability: str,
    payload_summary: Optional[Dict[str, Any]] = None,
) -> LedgerEventCreate:
    """Build a ledger event for agent invocation."""
    return build_ledger_event(LedgerEventOptions(
        request=request,
        response=response,
        resource_id=invocation_id,
        metadata={
            "agent_id": agent_id,
            "capability": capability,
            "payload_summary": payload_summary or {},
            "event_type": "agent:invoked",
        },
        policy_pack_name=MESH_POLICY_PACK["name"],
        policy_pack_version=MESH_POLICY_PACK["version"],
        request_context={
            "agent_id": agent_id,
            "capability": capability,
        },
        response_summary={
            "invoked": response.decision in (PolicyEffect.ALLOW, PolicyEffect.WARN, PolicyEffect.SHADOW_DENY),
        },
    ))


def build_agent_register_event(
    request: PolicyEvaluateRequest,
    response: PolicyEvaluateResponse,
    agent_id: str,
    agent_name: str,
    capabilities: List[str],
) -> LedgerEventCreate:
    """Build a ledger event for agent registration."""
    return build_ledger_event(LedgerEventOptions(
        request=request,
        response=response,
        resource_id=agent_id,
        metadata={
            "agent_name": agent_name,
            "capabilities": capabilities,
            "event_type": "agent:registered",
        },
        policy_pack_name=MESH_POLICY_PACK["name"],
        policy_pack_version=MESH_POLICY_PACK["version"],
    ))


# ============================================
# INFRA HELPERS
# ============================================

def build_operator_infra_event(
    request: PolicyEvaluateRequest,
    response: PolicyEvaluateResponse,
    operation_id: str,
    operation_type: str,
    target_service: str,
    details: Optional[Dict[str, Any]] = None,
) -> LedgerEventCreate:
    """Build a ledger event for infrastructure operations."""
    return build_ledger_event(LedgerEventOptions(
        request=request,
        response=response,
        resource_id=operation_id,
        metadata={
            "operation_type": operation_type,
            "target_service": target_service,
            "details": details or {},
            "event_type": f"operator:{operation_type}",
        },
        policy_pack_name=INFRA_POLICY_PACK["name"],
        policy_pack_version=INFRA_POLICY_PACK["version"],
        request_context={
            "operation_id": operation_id,
            "operation_type": operation_type,
            "target_service": target_service,
        },
        response_summary={
            "executed": response.decision in (PolicyEffect.ALLOW, PolicyEffect.WARN, PolicyEffect.SHADOW_DENY),
        },
    ))


def build_db_operation_event(
    request: PolicyEvaluateRequest,
    response: PolicyEvaluateResponse,
    operation_id: str,
    operation_type: str,  # migrate, backup, restore
    database: str,
    details: Optional[Dict[str, Any]] = None,
) -> LedgerEventCreate:
    """Build a ledger event for database operations."""
    return build_ledger_event(LedgerEventOptions(
        request=request,
        response=response,
        resource_id=operation_id,
        metadata={
            "operation_type": operation_type,
            "database": database,
            "details": details or {},
            "event_type": f"operator:{operation_type}",
        },
        policy_pack_name=INFRA_POLICY_PACK["name"],
        policy_pack_version=INFRA_POLICY_PACK["version"],
        request_context={
            "operation_id": operation_id,
            "database": database,
        },
        response_summary={
            "executed": response.decision in (PolicyEffect.ALLOW, PolicyEffect.WARN, PolicyEffect.SHADOW_DENY),
        },
    ))


# ============================================
# VALIDATION
# ============================================

def validate_ledger_event(event: LedgerEventCreate) -> None:
    """Validate a ledger event has all required fields.

    Raises ValueError if validation fails.
    """
    required_fields = [
        ("correlation_id", event.correlation_id),
        ("layer", event.layer),
        ("host", event.host),
        ("service", event.service),
        ("policy_scope", event.policy_scope),
        ("actor", event.actor),
        ("action", event.action),
        ("resource_type", event.resource_type),
        ("decision", event.decision),
        ("ledger_level", event.ledger_level),
    ]

    for field_name, value in required_fields:
        if value is None:
            raise ValueError(f"Missing required field: {field_name}")

    # Actor validation
    if not event.actor.user_id and not event.actor.agent_id:
        raise ValueError("Actor must have either user_id or agent_id")

    if not event.actor.role:
        raise ValueError("Actor must have a role")

    # Full fidelity validation
    if event.ledger_level == LedgerLevel.FULL:
        if not event.request_context:
            # Warning, not error - some full events may not have context
            pass

File: lib/ledger-builder.ts
/**
 * Ledger Event Builder for BlackRoad OS
 *
 * Spec for building ledger events from policy decisions.
 * This ensures all events have the required fields and
 * conform to the Amundson governance contract.
 *
 * @amundson 0.1.0
 * @governor alice.governor.v1
 * @operator alexa.operator.v1
 */

import type {
  PolicyEvaluateRequest,
  PolicyEvaluateResponse,
  LedgerEventCreate,
  LedgerLevel,
  Layer,
} from "./amundson-governance";

// ============================================
// CONSTANTS
// ============================================

/**
 * Current Amundson language version.
 * Must be included in all ledger events.
 */
export const LANGUAGE_VERSION = "0.1.0";

/**
 * Education policy pack metadata.
 * Update these when the pack version changes.
 */
export const EDU_POLICY_PACK = {
  name: "edu-policies",
  version: "1.0.0",
} as const;

// ============================================
// REQUIRED FIELDS SPEC
// ============================================

/**
 * REQUIRED FIELDS FOR STAGE 1 (Education Vertical)
 *
 * These fields MUST be present in every ledger event:
 *
 * IDENTITY:
 * - correlation_id: UUID linking this event to the request chain
 *
 * LOCATION:
 * - layer: "experience" for Education flows
 * - host: The domain (e.g., "edu.blackroad.io", "homework.blackroad.io")
 * - service: The service name (e.g., "edu-web")
 * - policy_scope: The policy namespace (e.g., "edu.*", "homework.*")
 *
 * ACTOR:
 * - actor.user_id: The user performing the action
 * - actor.role: Their role at time of action ("teacher" or "student")
 *
 * ACTION:
 * - action: The action attempted (e.g., "assignment:create")
 * - resource_type: The resource type (e.g., "assignment", "submission")
 * - resource_id: The specific resource ID (if applicable)
 *
 * DECISION:
 * - decision: The policy decision ("allow", "deny", "warn", "shadow_deny")
 * - policy_id: Which policy matched (null if default stance)
 * - policy_version: Version string from the policy
 *
 * METADATA:
 * - ledger_level: Must be >= required_ledger_level from policy response
 *
 * OPTIONAL BUT RECOMMENDED:
 * - asserted_facts: Facts the caller asserted
 * - fact_evidence: Evidence for those facts
 * - claims: Claims presented for evaluation
 * - metadata: Action-specific data (e.g., assignment title)
 *
 * CONDITIONAL (based on ledger_level):
 * - If ledger_level = "full":
 *   - request_context: Full request details
 *   - response_summary: Response details
 */

// ============================================
// BUILDER OPTIONS
// ============================================

export interface LedgerEventOptions {
  /** The original policy evaluation request */
  request: PolicyEvaluateRequest;

  /** The policy evaluation response */
  response: PolicyEvaluateResponse;

  /** The resource ID (if action was on a specific resource) */
  resourceId?: string | null;

  /** Additional metadata to include */
  metadata?: Record<string, unknown>;

  /** Override the ledger level (must be >= response.required_ledger_level) */
  ledgerLevelOverride?: LedgerLevel;

  /** Full request context (required if ledger_level = "full") */
  requestContext?: Record<string, unknown>;

  /** Response summary (required if ledger_level = "full") */
  responseSummary?: Record<string, unknown>;

  /** Intent ID if this is part of a multi-step flow */
  intentId?: string;

  /** Sequence number within the intent */
  sequenceNum?: number;
}

// ============================================
// LEDGER LEVEL PRECEDENCE
// ============================================

const LEDGER_LEVEL_RANK: Record<LedgerLevel, number> = {
  none: 0,
  decision: 1,
  action: 2,
  full: 3,
};

/**
 * Ensure the actual ledger level is >= required level.
 */
function resolveLevel(
  required: LedgerLevel,
  override?: LedgerLevel
): LedgerLevel {
  if (!override) return required;

  if (LEDGER_LEVEL_RANK[override] < LEDGER_LEVEL_RANK[required]) {
    console.warn(
      `Ledger level override "${override}" is below required "${required}". Using required.`
    );
    return required;
  }

  return override;
}

// ============================================
// LAYER RESOLUTION
// ============================================

/**
 * Resolve the layer from the host.
 */
function resolveLayer(host: string): Layer {
  if (host.includes("gov.") || host.includes(".systems")) {
    return "governance";
  }
  if (host.includes("mesh.") || host.includes("agents.")) {
    return "mesh";
  }
  if (host.includes("api.blackroad.io")) {
    return "gateway";
  }
  if (host.includes("db.") || host.includes("infra.") || host.includes("ops.")) {
    return "infra";
  }
  return "experience";
}

/**
 * Resolve the policy scope from the host.
 */
function resolvePolicyScope(host: string): string {
  if (host.startsWith("edu.")) return "edu.*";
  if (host.startsWith("homework.")) return "homework.*";
  if (host.startsWith("app.")) return "app.*";
  if (host.startsWith("gov.api.")) return "gov.api.*";
  if (host.startsWith("mesh.")) return "mesh.*";
  if (host.startsWith("agents.")) return "agents.*";
  if (host.startsWith("api.")) return "api.*";
  return "app.*";
}

// ============================================
// BUILDER FUNCTION
// ============================================

/**
 * Build a ledger event from a policy evaluation request/response pair.
 *
 * This is the canonical way to create ledger events. It ensures:
 * 1. All required fields are present
 * 2. Ledger level is >= required_ledger_level
 * 3. Location fields are consistent with the host
 * 4. Evidence from the request is preserved
 *
 * @example
 * // After evaluating policy:
 * const response = await evaluatePolicy(request);
 *
 * // Build and record the event:
 * const event = buildLedgerEvent({
 *   request,
 *   response,
 *   resourceId: "a123",
 *   metadata: { title: "Week 1 Homework" },
 * });
 *
 * await recordEvent(event);
 */
export function buildLedgerEvent(
  options: LedgerEventOptions
): LedgerEventCreate {
  const { request, response, resourceId, metadata, intentId, sequenceNum } =
    options;

  const host = request.context.request_metadata.host ?? "app.blackroad.io";
  const service = request.context.request_metadata.service ?? "unknown";
  const correlationId =
    request.context.request_metadata.correlation_id ?? crypto.randomUUID();

  const ledgerLevel = resolveLevel(
    response.required_ledger_level,
    options.ledgerLevelOverride
  );

  const event: LedgerEventCreate = {
    // Identity
    correlation_id: correlationId,
    intent_id: intentId ?? null,
    sequence_num: sequenceNum ?? 0,

    // Location
    layer: resolveLayer(host),
    host,
    service,
    policy_scope: resolvePolicyScope(host),

    // Actor
    actor: {
      user_id: request.subject.user_id ?? null,
      role: request.subject.role,
      agent_id: null,
      delegation_id: null,
    },

    // Action
    action: request.action,
    resource_type: request.resource.type,
    resource_id: resourceId ?? request.resource.id ?? null,

    // Decision
    decision: response.decision,
    policy_id: response.policy_id ?? null,
    policy_version: response.policy_version ?? null,

    // Evidence
    asserted_facts: request.context.asserted_facts,
    fact_evidence: request.context.fact_evidence,
    claims: request.context.claims,

    // Metadata
    ledger_level: ledgerLevel,
    metadata: {
      ...metadata,
      language_version: LANGUAGE_VERSION,
      policy_pack: EDU_POLICY_PACK.name,
      policy_pack_version: EDU_POLICY_PACK.version,
    },

    // Full fidelity fields (only if ledger_level = "full")
    request_context: ledgerLevel === "full" ? options.requestContext : null,
    response_summary: ledgerLevel === "full" ? options.responseSummary : null,
  };

  return event;
}

// ============================================
// CONVENIENCE: EDUCATION VERTICAL HELPERS
// ============================================

/**
 * Build a ledger event for teacher assignment creation.
 */
export function buildAssignmentCreateEvent(
  request: PolicyEvaluateRequest,
  response: PolicyEvaluateResponse,
  assignmentId: string,
  assignmentTitle: string
): LedgerEventCreate {
  return buildLedgerEvent({
    request,
    response,
    resourceId: assignmentId,
    metadata: {
      title: assignmentTitle,
      event_type: "assignment:created",
    },
  });
}

/**
 * Build a ledger event for student submission.
 */
export function buildSubmissionEvent(
  request: PolicyEvaluateRequest,
  response: PolicyEvaluateResponse,
  submissionId: string,
  assignmentId: string
): LedgerEventCreate {
  return buildLedgerEvent({
    request,
    response,
    resourceId: submissionId,
    metadata: {
      assignment_id: assignmentId,
      event_type: "submission:submitted",
    },
  });
}

/**
 * Build a ledger event for teacher review.
 */
export function buildReviewEvent(
  request: PolicyEvaluateRequest,
  response: PolicyEvaluateResponse,
  submissionId: string,
  reviewResult: "approved" | "returned" | "flagged"
): LedgerEventCreate {
  return buildLedgerEvent({
    request,
    response,
    resourceId: submissionId,
    metadata: {
      review_result: reviewResult,
      event_type: "submission:reviewed",
    },
  });
}

// ============================================
// VALIDATION
// ============================================

/**
 * Validate a ledger event has all required fields for Stage 1.
 * Throws if validation fails.
 */
export function validateLedgerEvent(event: LedgerEventCreate): void {
  const required: (keyof LedgerEventCreate)[] = [
    "correlation_id",
    "layer",
    "host",
    "service",
    "policy_scope",
    "actor",
    "action",
    "resource_type",
    "decision",
    "ledger_level",
  ];

  for (const field of required) {
    if (event[field] === undefined || event[field] === null) {
      throw new Error(`Missing required field: ${field}`);
    }
  }

  if (!event.actor.user_id && !event.actor.agent_id) {
    throw new Error("Actor must have either user_id or agent_id");
  }

  if (!event.actor.role) {
    throw new Error("Actor must have a role");
  }

  // Validate full fidelity requirements
  if (event.ledger_level === "full") {
    if (!event.request_context) {
      console.warn(
        "ledger_level=full but request_context is missing. Consider adding it."
      );
    }
  }
}
